src/demikernel/dbuf.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+//==============================================================================
+// Imports
+//==============================================================================
+
+use ::runtime::{
+    fail::Fail,
+    memory::Buffer,
+};
+use ::std::{
+    fmt,
+    fmt::Debug,
+    ops::{
+        Deref,
+        DerefMut,
+    },
+    slice,
+    sync::Arc,
+};
+
+//==============================================================================
+// Structures
+//==============================================================================
+
+/// Data Buffer.
+#[derive(Clone)]
+pub struct DataBuffer {
+    /// Underlying data.
+    data: Option<Arc<[u8]>>,
+
+    /// Data offset.
+    offset: usize,
+
+    /// Data length
+    len: usize,
+}
+
```

Comment:
Just a comment, not something to change now.

We should think about having various additional fields:
- an identifier for the memory allocator the buffer space came from.
- one that "remembers" the original length of the data allocation (i.e. before any adjust or trim), as it might be useful in the future.
- next and prev pointers for putting on lists w/o extra allocations.
- an "extra" field that each LibOs can use for whatever it wants (we should have space for another 64 bit field in this struct before we exceed the x86 cache line size).

Reply:
Sure. I created issue https://github.com/demikernel/demikernel/issues/106 to keep track of this.